### ISSUE TITLE
FIPS compliant images

### ISSUE THREAD

**Issue by faisal-memon** [spiffe/spire#4009]
It would be great to build FIPS compliant images for American government agencies and contractors. The scratch image is a good start. We need to meet other requirement like working with FIPS OpenSSL.

**Comment by edwbuck** [comment]
After a bit of research, my recommendation is that we offer a custom WolfSSL key manager, and update the documentation to comment of FIPS compliance; but, this will not be enough to be truly FIPS compliant, as we would have to also move all of the signing validation logic into either the key manager interface or a new interface, again backed by WolfSSL libraries.  The documentation would have to note this in detail.

The alternative recommendation is that we simply maintain no claims of FIPS compliance.  Oddly enough FIPS compliance is not a controlled category.  It is a non-controlled claim that your software uses FIPS Certified cryptographic modules, aka libraries and / or hardware where the algorithms have been validated by the NIST to be FIPS Verified, and then additionally issued a certificate stating that specific release / build / configuration is FIPS Certified.  Any changes, including bug fixes, invalids the verification and certification.

Issuing images that advertise FIPS compliance is likely the wrong approach.  WolfSSL is currently the only provider of FIPS Certified libraries, and the informal "FIPS Compliant" statement is based on using only FIPS Certified components.  Software updates in dependent libraries would not advertise if the new updates have a registered NIST FIPS Compliance document, or even if they were FIPS Validated.  By making FIPS compliance a configuration issue the SPIRE brand is protected when deployed in ways that use the wrong SSL libraries, or even the correct SSL libraries that lack NIST Certification or lack NIST Certification and NIST Validation.

Why the focus on WolfSSL?  Their FIPS-OpenSSL offering is advertised as the only FIPS-2 compliant module in existence.  Currently FIPS-2 is deprecated, replaced by FIPS-3.  While FIPS-3 started in 2019, it only received its first submission for Validation in 2020 and to date no modules have been issued FIPS-3 Certification.  

Criticism of the FIPS program is fierce.

I would ask the core maintainers to choose between the two options, and then we can decide on the answer to the question "Does SPIRE provide FIPS compliance?"  being one of "FIPS compliance is dependent on the deployment configuration of SPIRE, where one needs to configure an maintain a valid configuration against a FIPS compliant key manager" or "No, SPIRE leverages the latest OpenSSL libraries to ensure the fastest path for bug and security fixes, and the latest OpenSSL libraries choose to value speedy repair of security issues over the delays that FIPS Certification by NIST require."

Once we have the guidance, we should update / create a "frequently asked questions" section of the SPIRE documentation to expose the decision.

**Comment by v0lkan** [comment]
Related:

I recently did a bunch of FIPS-related work in Aegis ( https://github.com/shieldworks/aegis/issues?q=is%3Aissue+FIPS ) 

What I did was
* swap the encryption algorithm with RSA (with an optional flag; the default algorithm is still Age Encryption)
* use a fips-compliant go runtime during the build phase (go has an experimental `boringcrypto` flag that uses FIPS-compliant crypto

And that was pretty much it.

Since SPIRE is a single binary, maintaining compliance should not take too much work.

--

btw, +1 with WolfSSL.

**Comment by v0lkan** [comment]
@faisal-memon 

What are the rough steps if we wanted to create a PR to ensure FIPS compliance. Can you give a step-by-step instruction, especially for newcomers?

**Comment by v0lkan** [comment]
Related:

> By making FIPS compliance a configuration issue the SPIRE brand is protected when deployed in ways that use the wrong SSL libraries, or even the correct SSL libraries that lack NIST Certification or lack NIST Certification and NIST Validation.

That’s how Aegis considers FIPS-compliance by the way:

Quote from its documentation:

```
AEGIS_SAFE_FIPS_COMPLIANT

AEGIS_SAFE_FIPS_COMPLIANT is required for Aegis Safe to run in FIPS-compliant mode. Defaults to "false", which means Aegis Safe will run in non-FIPS-compliant mode. Setting it to "true" will make Aegis Safe run in FIPS-compliant mode.

Note that this is not a guarantee that Aegis Safe will actually run in FIPS compliant mode, as it depends on the underlying base image.

If you are using one of the official FIPS-complaint Aegis Docker images, then it will be FIPS-compliant.

As a FIPS-compliant base image you can choose from the following:

[aegishub/aegis-ist-fips-safe](https://hub.docker.com/repository/docker/aegishub/aegis-ist-fips-safe/general) (using a Distroless base)
[aegishub/aegis-photon-fips-safe](https://hub.docker.com/repository/docker/aegishub/aegis-photon-fips-safe/general) (using VMware Photon OS as a base)
```

**Comment by Loadingname91** [comment]
> @faisal-memon
> 
> What are the rough steps if we wanted to create a PR to ensure FIPS compliance. Can you give a step-by-step instruction, especially for newcomers?

Hey @faisal-memon , did you find a chance to prepare this ? I was also looking into FIPS compliant binary and was wondering if you got any updates which can help me

**Comment by shashankram** [comment]
> I recently did a bunch of FIPS-related work in Aegis ( https://github.com/shieldworks/aegis/issues?q=is%3Aissue+FIPS )
> 
> What I did was
> 
>     * swap the encryption algorithm with RSA (with an optional flag; the default algorithm is still Age Encryption)
> 
>     * use a fips-compliant go runtime during the build phase (go has an experimental `boringcrypto` flag that uses FIPS-compliant crypto
> 
> 
> Sharing here, in case it might help.
> 
> And that was pretty much it.
> 
> Since SPIRE is a single binary, maintaining compliance should not take too much work.
> 

This seems reasonable to me. @v0lkan would you be open to discussing this in an upcoming contributor sync (happens every Tuesday at 11.30 am pacific time).
FIPS compliance is a big deal in restricted environments.

**Comment by v0lkan** [comment]
@shashankram Sure, I am busy for a week or two, but I can jump into the earliest contributor sync I can make. — I’ll let here know.

**Comment by shashankram** [comment]
@amartinezfayo is this expected to land in 1.9?

**Comment by amartinezfayo** [comment]
@shashankram We haven't assigned a milestone to this yet.
We have discussed this and the SPIRE project is not in a position that could publish FIPS compliant images. Rather, we think that we could make it easier to have SPIRE built (and probably configured) in a manner that adheres to the standards that FIPS 140-2 establishes.
To achieve this, users will need to build SPIRE with a FIPS validated cryptographic library. One option could be to build SPIRE with [BoringSSL](https://boringssl.googlesource.com/boringssl/). BoringSSL as a whole is not FIPS validated, but a specific core library called [BoringCrypto](https://boringssl.googlesource.com/boringssl/+/master/crypto/fipsmodule/FIPS.md) is. We can leverage the use of [BUILD_GOEXPERIMENT=boringcrypto](https://pkg.go.dev/internal/goexperiment@go1.19) to enable integrating BoringCrypto in the Go toolchain when building SPIRE.

**Comment by shashankram** [comment]
> @shashankram We haven't assigned a milestone to this yet. We have discussed this and the SPIRE project is not in a position that could publish FIPS compliant images. Rather, we think that we could make it easier to have SPIRE built (and probably configured) in a manner that adheres to the standards that FIPS 140-2 establishes. To achieve this, users will need to build SPIRE with a FIPS validated cryptographic library. One option could be to build SPIRE with [BoringSSL](https://boringssl.googlesource.com/boringssl/). BoringSSL as a whole is not FIPS validated, but a specific core library called [BoringCrypto](https://boringssl.googlesource.com/boringssl/+/master/crypto/fipsmodule/FIPS.md) is. We can leverage the use of [BUILD_GOEXPERIMENT=boringcrypto](https://pkg.go.dev/internal/goexperiment@go1.19) to enable integrating BoringCrypto in the Go toolchain when building SPIRE.

@amartinezfayo that sounds reasonable to me. Curious if this is scoped for the upcoming release.

**Comment by nweisenauer-sap** [comment]
Any updates on this? FIPS compliance is also an issue for us.

**Comment by nikotih** [comment]
Hi @amartinezfayo, can we revisit this topic again? 

we were evaluating spire fips compliance with new golang fips flags
https://go.dev/doc/security/fips140

tldr: It can detect during build and panic during runtime
main challenge we discovered seems to be the use of sha1 for key signatures (suppressed gosec in multiple places in spire code)

can we have any mitigates for those? or this would be a breaking change?

**Comment by amartinezfayo** [comment]
Thank you @nikotih for reaching out and identifying the SHA-1 usage as a concern for FIPS compliance.
I've identified the specific places where SHA-1 is used in SPIRE (running grep -R '"crypto/sha1"' .):

- ./pkg/server/plugin/nodeattestor/tpmdevid/selectors.go
- ./pkg/server/plugin/keymanager/gcpkms/gcpkms.go
- ./pkg/agent/plugin/svidstore/gcpsecretmanager/gcloud.go
- ./pkg/common/plugin/x509pop/x509pop.go
- ./pkg/common/x509util/keyid.go

Any changes affecting node attestation or certificate key IDs would require careful migration planning to avoid breaking existing deployments. Other references may be easier to migrate to SHA-256.

A complete removal of SHA-1 would indeed be breaking.

Beyond FIPS compliance, moving away from SHA-1 is something that I personally think it's worth considering.
If you're interested in pursuing this, could you please open a dedicated issue? This would allow us to properly scope the work required for each usage location and track community interest and potential contributions.
